### PR TITLE
Goa 2815 go ancestor service requirements

### PR DESCRIPTION
**Pull request by TonyWardell**
Graph data structure for ontology returned for start nodes, using stop nodes if supplied, relations if supplied.

### CODE REVIEW

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/SubGraphCalculator.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ * @author Tony Wardell
+ * Date: 19/06/2017
+ * Time: 10:11
+ * Created with IntelliJ IDEA.
+ */
+class SubGraphCalculator {
+
+    private static Logger LOGGER = LoggerFactory.getLogger(SubGraphCalculator.class);
+
+    static Trampoline<AncestorGraph> createTrampoline(AncestorGraphRequest request, AncestorGraph<String> ancestorGraph,
```
Comment (eddturner) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/SubGraphCalculator.java]: I think this should have a clearer method name to describe what it is doing.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/controller/OBOController.java [hunk 2]
```diff
@@ -93,7 +95,8 @@ public OBOController(OntologyService<T> ontologyService,
             OBOControllerValidationHelper oboControllerValidationHelper,
             OntologyRestConfig.OntologyPagingConfig ontologyPagingConfig,
             OntologyType ontologyType,
-            HttpHeadersProvider httpHeadersProvider) {
+            HttpHeadersProvider httpHeadersProvider,
+            List<OntologyRelationType> allowedGraphTraversalTypes) {
```
Comment (TonyWardell) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/controller/OBOController.java]: Created OntologySpecifier

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/SubGraphCalculator.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+/**
+ * Encapsulates the calculation of a edges for a single vertex, and the creation of further instances for vertexes
+ * attached to those edges.
+ * @author Tony Wardell
+ * Date: 19/06/2017
+ * Time: 10:11
+ * Created with IntelliJ IDEA.
+ */
+class SubGraphCalculator {
```
Comment (TonyWardell) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/SubGraphCalculator.java]: Like you say, I probably wouldn't do it this way again, but sincerely I don't think it's so bad as to be worth re-coding - it does encapsulate the sub-graph calculation nicely, and makes the testing easier, as the reference to the graph can be passed in and is much easer to mock than building up the values in OntologyGraph for a test.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/SubGraphCalculator.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        Preconditions.checkArgument(Objects.nonNull(ontologyGraphTraversal), "SubGraphCalculator#createTrampoline cannot accept an " +
+                "argument for ontologyGraphTraversal that is null");
+
+        String target = request.targetVertices.pollFirst();
+        if (Objects.nonNull(target)) {
+            if (ancestorGraph.vertices.add(target)) {
+
+                //if target is not a stop node look for parents
+                if (!request.stopVertices.contains(target)) {
+
```
Comment (TonyWardell) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/SubGraphCalculator.java]: Done.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/SubGraphCalculator.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                        return Optional.of(createTrampoline(request,
+                                                            ancestorGraph,
+                                                            ontologyGraphTraversal));
+                    }
+                };
+            }
+
+        }
+        return new Trampoline<AncestorGraph>() {
+            public AncestorGraph getValue() { return ancestorGraph; }
```
Comment (TonyWardell) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/SubGraphCalculator.java]: Done.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/Trampoline.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package uk.ac.ebi.quickgo.ontology.traversal;
+
+import java.util.Optional;
+
+/**
+ * An alternative to recursion.
```
Comment (TonyWardell) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/Trampoline.java]: Done.
Wish we had this for all of spring.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/Trampoline.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ * <p>
+ * https://stackoverflow.com/questions/32685660/achieving-stackless-recursion-in-java-8
+ * <p>
+ * Created by Tony Wardell on 17-Jun-17.
+ */
+public class Trampoline<T> {
```
Comment (TonyWardell) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/Trampoline.java]: Added.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/Trampoline.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    public T getValue() {
+        throw new RuntimeException("Not implemented");
```
Comment (TonyWardell) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/Trampoline.java]: Done.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/Trampoline.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    public Optional<Trampoline<T>> nextTrampoline() {
+        return Optional.empty();
+    }
+
+    final T compute() {
```
Comment (TonyWardell) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/Trampoline.java]: Please explain.

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/controller/OBOControllerIT.java [hunk 10]
```diff
@@ -88,7 +92,7 @@
     private String validIdsCSV;
     private List<String> validIdShortList;
     private List<String> validIdList;
-    private List<OntologyRelationship> relationships;
+    private List<OntologyRelationship>  relationships;
```
Comment (TonyWardell) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/controller/OBOControllerIT.java]: Done.

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/controller/validation/OBOControllerValidationHelperImplTest.java [hunk 11]
```diff
@@ -39,7 +39,7 @@ public void setUp() {
     @Test
```
Comment (TonyWardell) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/controller/validation/OBOControllerValidationHelperImplTest.java]: Done.

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/service/OntologyServiceImplTest.java [hunk 12]
```diff
@@ -116,6 +116,13 @@ private GOTerm createGOTerm(String id) {
             return term;
         }
 
+        private GOTerm createGOTerm(String id, String name) {
```
Comment (TonyWardell) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/service/OntologyServiceImplTest.java]: Done.

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/service/OntologyServiceImplTest.java [hunk 13]
```diff
@@ -293,7 +300,7 @@ public void findsAnnotationGuideLinesInfoForGoIdentifier() {
 
         @Test
         public void findsEmptyAncestorsForMissingTerm() {
-            Set<String> ids = idsViaOntologyService("GO:0000001").stream().collect(Collectors.toSet());
+            Set<String> ids = new HashSet<>(idsViaOntologyService("GO:0000001"));
```
Comment (TonyWardell) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/service/OntologyServiceImplTest.java]: Intellij was complaining about the previous code.

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/service/OntologyServiceImplTest.java [hunk 14]
```diff
@@ -766,6 +847,6 @@ public void findsPathsForTermWithRelation() {
     }
 
     private static <ItemType> List<ItemType> asList(Collection<ItemType> items) {
-        return items.stream().collect(Collectors.toList());
+        return new ArrayList<>(items);
```
Comment (TonyWardell) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/service/OntologyServiceImplTest.java]: Idea recommendation.

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/service/OntologyServiceImplTest.java [hunk 15]
```diff
@@ -452,6 +460,79 @@ public void findsPathsForTermWithRelation() {
         private List<String> idsViaOntologyService(String... ids) {
             return goOntologyService.buildIdList(Arrays.asList(ids));
         }
+
+        //SUB-GRAPH
+
```
Comment (TonyWardell) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/service/OntologyServiceImplTest.java]: Done.

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/service/OntologyServiceImplTest.java [hunk 16]
```diff
@@ -452,6 +460,79 @@ public void findsPathsForTermWithRelation() {
...
+        @Test
+        public void populatedAncestorGraphForSelectedTerm() {
```
Comment (TonyWardell) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/service/OntologyServiceImplTest.java]: Done.

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/service/OntologyServiceImplTest.java [hunk 17]
```diff
@@ -452,6 +460,79 @@ public void findsPathsForTermWithRelation() {
...
+            String child = "GO:0000001";
```
Comment (TonyWardell) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/service/OntologyServiceImplTest.java]: Done.

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/traversal/SubGraphCalculatorTest.java [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+import static org.mockito.Mockito.when;
+import static uk.ac.ebi.quickgo.ontology.model.OntologyRelationType.IS_A;
+
+/**
+ * @author Tony Wardell
+ * Date: 21/06/2017
+ * Time: 08:53
+ * Created with IntelliJ IDEA.
+ */
+public class SubGraphCalculatorTest {
```
Comment (TonyWardell) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/traversal/SubGraphCalculatorTest.java]: Done.

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraphTest.java [hunk 19]
```diff
@@ -743,4 +742,272 @@ public void doesNotFinGrandParentOfChildWhenSearchingForParents() throws Excepti
...
+
+            assertThat(ancestorGraph.vertices, hasSize(3));
+            assertThat(ancestorGraph.vertices, containsInAnyOrder(pyrophosphataseActivity.id, cyclaseActivity.id,
+                                                                  catalyticActivity.id));
+            assertThat(ancestorGraph.edges, hasSize(2));
+            assertThat(ancestorGraph.edges, containsInAnyOrder(toAE(py_IA_cy), toAE(cy_IA_ca)));
+        }
+
+        @Test
+        public void findSubGraphWhereMultipleRouteOfSameType() {
```
Comment (TonyWardell) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraphTest.java]: Done.

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraphTest.java [hunk 20]
```diff
@@ -743,4 +742,272 @@ public void doesNotFinGrandParentOfChildWhenSearchingForParents() throws Excepti
...
+            AncestorGraph<String> ancestorGraph = og.subGraph(startingVertices, stopVertices, relations);
+
+            assertThat(ancestorGraph.vertices, hasSize(2));
+            assertThat(ancestorGraph.vertices, containsInAnyOrder(pyrophosphataseActivity.id, molecularFunction.id));
+            assertThat(ancestorGraph.edges, hasSize(1));
+            assertThat(ancestorGraph.edges, containsInAnyOrder(toAE(py_OI_mf)));
+        }
+
+        @Test
+        public void UnrelatedVerticesIgnored() {
```
Comment (TonyWardell) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraphTest.java]: Done.

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraphTest.java [hunk 21]
```diff
@@ -743,4 +742,272 @@ public void doesNotFinGrandParentOfChildWhenSearchingForParents() throws Excepti
...
+            final HashSet<String> startingVertices = new HashSet<>(singletonList("GO:XXXXXX"));
+
+            AncestorGraph<String> ancestorGraph = og.subGraph(startingVertices, stopVertices, relations);
+
+            assertThat(ancestorGraph.vertices, hasSize(0));
+            assertThat(ancestorGraph.edges, hasSize(0));
+        }
+
+        @Test
+        public void canFindRootNodeOnly() {
```
Comment (TonyWardell) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraphTest.java]: Removed canFindRootNodeOnly

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraphTest.java [hunk 22]
```diff
@@ -743,4 +742,272 @@ public void doesNotFinGrandParentOfChildWhenSearchingForParents() throws Excepti
...
+
+            AncestorGraph<String> ancestorGraph = og.subGraph(startingVertices, stopVertices, relations);
+
+            assertThat(ancestorGraph.vertices, hasSize(1));
+            assertThat(ancestorGraph.vertices, containsInAnyOrder(molecularFunction.id));
+            assertThat(ancestorGraph.edges, hasSize(0));
+        }
+
+        @Test
+        public void rootNodeCanBeStopNodeCanBeOnlyNode() {
```
Comment (TonyWardell) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraphTest.java]: Renamed.

File: ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraphTest.java [hunk 23]
```diff
@@ -743,4 +742,272 @@ public void doesNotFinGrandParentOfChildWhenSearchingForParents() throws Excepti
...
+
+            AncestorGraph<String> ancestorGraph = og.subGraph(startingVertices, stopVertices, relations);
+
+            assertThat(ancestorGraph.vertices, hasSize(1));
+            assertThat(ancestorGraph.vertices, containsInAnyOrder(molecularFunction.id));
+            assertThat(ancestorGraph.edges, hasSize(0));
+        }
+
+        @Test
+        public void cyclicalOntologyDoNotHitCycleAsHitStopNodeFirst() {
```
Comment (TonyWardell) [inline ontology-rest/src/test/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraphTest.java]: cyclicalOntologyCanBeCalculatedWithoutEnteringInfiniteLooping

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/model/graph/AncestorGraphRequest.java [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * Data structure that describes the requirements for a sub-graph from an ontology.
+ * @author Tony Wardell
+ * Date: 21/06/2017
+ * Time: 09:03
+ * Created with IntelliJ IDEA.
+ */
+public class AncestorGraphRequest {
+
+    public final Deque<String> targetVertices;
```
Comment (TonyWardell) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/model/graph/AncestorGraphRequest.java]: Yeah, I always knew Queue was sufficient but I liked the idea nodes would be processed in the order they were found, despite it being complete unnecessary.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java [hunk 25]
```diff
@@ -74,29 +92,24 @@ public void addRelationships(Collection<? extends OntologyRelationship> relation
 
         Preconditions.checkArgument(!isNullOrEmpty(startingVertices), "Starting vertices cannot be null/empty.");
         Preconditions.checkArgument(!isNullOrEmpty(endingVertices), "Ending vertices cannot be null/empty.");
+        errorIfAllStartingVerticesAreEndingVertices(startingVertices, endingVertices);
```
Comment (TonyWardell) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java]: I didn't change the behaviour - just moved the errorIf.. code into its own method.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java [hunk 26]
```diff
@@ -161,11 +173,37 @@ public void addRelationships(Collection<? extends OntologyRelationship> relation
         Preconditions.checkArgument(topVertex != null && !topVertex.trim().isEmpty(),
                 "Top vertex cannot be null or empty");
 
-        Set<OntologyRelationType> relationsSet = createRelevantRelationsSet(relations);
+        Set<OntologyRelationType> relationsSet = createRelevantRelationSet(relations);
```
Comment (TonyWardell) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java]: Agreed, don't know what I was thinking there.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java [hunk 27]
```diff
@@ -161,11 +173,37 @@ public void addRelationships(Collection<? extends OntologyRelationship> relation
...
 
         return ontology.incomingEdgesOf(topVertex).stream()
                 .filter(ontologyRelationship -> relationsSet.contains(ontologyRelationship.relationship))
-                .collect(Collectors.toSet());
+                .collect(toSet());
+    }
+
+    @Override public AncestorGraph<String> subGraph(Set<String> startVertices, Set<String> stopVertices,
```
Comment (TonyWardell) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java]: Done.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java [hunk 28]
```diff
@@ -161,11 +173,37 @@ public void addRelationships(Collection<? extends OntologyRelationship> relation
...
+            OntologyRelationType... relations) {
+        Preconditions.checkArgument(!isNullOrEmpty(startVertices), "Starting vertices cannot be null/empty.");
+        // Valid relationship list is ontology dependent and therefore must be supplied.
+        // We can't use the full list as there maybe edges with relationships that should not be displayed in the
+        // subgraph
+        Preconditions.checkArgument(Objects.nonNull(relations) && relations.length > 0, "Relations cannot be null");
+        AncestorGraph<String> ancestorGraph = new AncestorGraph<>(new HashSet<>(), new HashSet<>());
```
Comment (TonyWardell) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java]: But you would have a default type for AncestorGraph then, which I think isn't right.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java [hunk 29]
```diff
@@ -161,11 +173,37 @@ public void addRelationships(Collection<? extends OntologyRelationship> relation
...
+        Deque<String> targetVertices = buildTargetVertices(startVertices);
+        if(!targetVertices.isEmpty()){
```
Comment (TonyWardell) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java]: Done.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java [hunk 30]
```diff
@@ -161,11 +173,37 @@ public void addRelationships(Collection<? extends OntologyRelationship> relation
...
+            stopVertices.addAll(STOP_NODES);
+            OntologyRelationType[] targetRelations = OntologyRelationType.relevantRelations(relations);
+            AncestorGraphRequest request = new AncestorGraphRequest(targetVertices, stopVertices, targetRelations);
+            SubGraphCalculator.createTrampoline(request, ancestorGraph, this).compute();
```
Comment (TonyWardell) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraph.java]: Should be clear now.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraphTraversal.java [hunk 31]
```diff
@@ -75,4 +76,18 @@
...
     Set<OntologyRelationship> children(String topVertex, OntologyRelationType... relations);
-}
\ No newline at end of file
+
+    /**
+     * Find the sub-graph between two sets of vertices in a graph, navigable via
+     * a specified set of relations.
+     *
+     * @param startVertices the base vertices which are the lowest level of the sub-graph
+     * @param stopVertices the ending vertices beyond which Ontology vertices and edges are not returned. If this
```
Comment (TonyWardell) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/OntologyGraphTraversal.java]: Done.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/SubGraphCalculator.java (hunk 3 above)
Comment (eddturner) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/traversal/SubGraphCalculator.java]: Agreed.

File: ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/model/graph/AncestorGraphRequest.java (hunk 24 above)
Comment (eddturner) [inline ontology-rest/src/main/java/uk/ac/ebi/quickgo/ontology/model/graph/AncestorGraphRequest.java]: So if unnecessary shouldn't it be a Queue? (I'd go for the one with the best performance, given how it is being used)

**Comment by eddturner** [comment]
@TonyWardell, **please do not** merge this to master until @xwatkins and co. have agreed that the service fulfils their needs (e.g., enables the building of a prototype). Since this feature is not a required part of the new QGO production, it is not essential (@SangyaP can confirm).

**Pull request merged**